### DISCUSSION TITLE
How to use the new stream.xlsx.WorkbookReader()

### DISCUSSION THREAD

**Discussion by alangunning** [exceljs/exceljs#2517] [Q&A]
Hi,

Is there any example of how to use the new stream.xlsx.WorkbookReader() where the input is a file stream? I have been unable to detect 'entry' or 'worksheet' events only can detect 'finished'.

Thanks,
Alan

**Comment by alangunning** [comment]
Hi, 

I started to receive the worksheet and row events once I set the options correctly. 

Accessing each row I can now get the cell values for each row as seen below. However, I am still not receiving the 'shared-strings' event. Can you please tell me how I can access the the actual values of shared Strings ? 

Thanks,
Alan

 row.values [ , { sharedString: 2 }, { sharedString: 3 } ]
 row.model { cells: 
   [ { address: 'A2', type: 7, value: [Object], style: {} },
     { address: 'B2', type: 7, value: [Object], style: {} } ],
  number: 2,
  min: 1,
  max: 2,
  height: 12.8,
  style: {} }

**Comment by bluedge** [comment]
Hi, Could you please show us the code you use to get this output from WorkbookReader() ?
Thanks

**Comment by alangunning** [comment]
@bluedge I don't have access to the exact code that I used, nor have I been following any of the subsequent updates to this library. With that in mind, this code is untested and was only used for prototype testing

```
    var workbook = new exceljs.stream.xlsx.WorkbookReader();
    var options = {
        entries: "emit",
        sharedStrings: "cache",
        styles: "emit",
        hyperlinks: "emit",
        worksheets: "emit"
    };
    workbook.on('error', function (error) {
            console.log('An error occurred while writing reading excel', error);        
    });

    workbook.on('entry', function (entry) {
        console.log("entry", entry);
    });

    workbook.on('shared-string', function (index, text) {
        console.log("index:", index, "text:", text);
    });

    workbook.on('worksheet', function (worksheet) {
            console.log("worksheet", worksheet.name);           
            worksheet.on('row', function (row) {
                console.log(" row.values", row.values);
                console.log(" row.model", row.model);
            });

            worksheet.on('close', function () {
                console.log("worksheet close");         
            });

            worksheet.on('finished', function () {
                console.log("worksheet finished");          
            });
    });

    workbook.on('finished', function () {
            console.log("finished");
    });

    workbook.on('close', function () {
            console.log("close");
    });

      workbook.read(readStream, options);
```

**Comment by bluedge** [comment]
Thanks a lot for your answer Alan !!!
I have tried to stream an .xlsx file (readStream) unfortunately it triggers an error when the library emit for 'shared-string'. I have tried to debug the code but it is yet hard to read and follow.
In case It is to someone interest I drop the error log below:
`nhandled rejection TypeError: sharedStrings.getString is not a function
    at SAXStream.<anonymous> (/usr/local/bin/www/mailReporter/node_modules/exceljs/lib/stream/xlsx/worksheet-reader.js:282:50)
    at emitOne (events.js:77:13)
    at SAXStream.emit (events.js:169:7)
    at Object.me._parser.(anonymous function) [as onclosetag] (/usr/local/bin/www/mailReporter/node_modules/exceljs/node_modules/sax/lib/sax.js:245:15)
    at emit (/usr/local/bin/www/mailReporter/node_modules/exceljs/node_modules/sax/lib/sax.js:615:33)
    at emitNode (/usr/local/bin/www/mailReporter/node_modules/exceljs/node_modules/sax/lib/sax.js:620:3)
    at closeTag (/usr/local/bin/www/mailReporter/node_modules/exceljs/node_modules/sax/lib/sax.js:861:5)
    at Object.write (/usr/local/bin/www/mailReporter/node_modules/exceljs/node_modules/sax/lib/sax.js:1294:29)
    at SAXStream.write (/usr/local/bin/www/mailReporter/node_modules/exceljs/node_modules/sax/lib/sax.js:227:16)
    at /usr/local/bin/www/mailReporter/node_modules/exceljs/lib/utils/flow-control.js:117:23
    at Array.forEach (native)
    at Function._.each._.forEach (/usr/local/bin/www/mailReporter/node_modules/exceljs/node_modules/underscore/underscore.js:78:11)
    at utils.inherits._pipe (/usr/local/bin/www/mailReporter/node_modules/exceljs/lib/utils/flow-control.js:114:7)
    at /usr/local/bin/www/mailReporter/node_modules/exceljs/lib/utils/flow-control.js:173:25
    at tryCatcher (/usr/local/bin/www/mailReporter/node_modules/exceljs/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/usr/local/bin/www/mailReporter/node_modules/exceljs/node_modules/bluebird/js/release/promise.js:503:31)`

**Comment by guyonroche** [comment]
Hi work on workbook reader is paused - there are memory issues with the implementation that I'm not happy with so I haven't properly released it yet. As always, when time allows I hope to work on it.

**Comment by gruentee** [comment]
@guyonroche 
Thanks for your work! I would really appreciate this feature being implemented.

**Comment by rfratto** [comment]
Hi, is there an update on this? What is still required to complete the WorkbookReader?

**Comment by Siemienik** [comment]
is it still actual?

**Comment by sumbad** [comment]
@Siemienik yes, it is still actual.

**Comment by colinroemer** [comment]
`workbook.on('worksheet', function (worksheet) {
     worksheet.on('row', function (row) {
    });
});`

is it possible to pass the row data into a writeable stream here?

For example, I am tasked with streaming an XLSX file from S3 through excelsjs and then streaming the data back to a new bucket in S3 as a CSV.  I understand this is an interesting ask but is a requirement for a story that I am working on.

**Comment by MichaelWalker-git** [comment]
```
    const Bucket = event.Records[0].s3.bucket.name;
    const Key = decodeURIComponent(event.Records[0].s3.object.key.replace(/\+/g, ' '));
    // Retrieve the object
    const data = await s3.getObject({ Bucket, Key }).createReadStream();

    const workbookReader = new ExcelJS.stream.xlsx.WorkbookReader(data);
    for await (const worksheetReader of workbookReader) {
        for await (const row of worksheetReader) {
            console.log("Raw text:\n" + row.values);
        }
    }

```

I get all the row values now from my S3 to my lambda

**Comment by aonamrata** [comment]
Do you know how i can get styles as well? I am using 
```
// templateBody is Buffer from S3 xlsx file
 const tempWorkbook = new Workbook();
 await tempWorkbook.xlsx.load(templateBody);
```
But this wont work with `WorkbookReader` if i am loading a very large file.

**Reply by vandungday** [reply]
I have issue too

**Reply by vandungday** [reply]
Are you fix it  ?

**Reply by aonamrata** [reply]
No, i just used a copy of a blank template file and added data to it.

**Reply by vandungday** [reply]
you can share code for me ?

I think you use workbook load 
and copy to stream workbook yet ?

**Reply by aonamrata** [reply]
I think this was what we ended up doing 
```
    const tempFilePath = `/tmp/${Date.now()}`;
        const options = {
            filename: tempFilePath,
            useStyles: true,
            useSharedStrings: true,
        };

    const combinedWorkbook = new stream.xlsx.WorkbookWriter(options);
    const templateBody = await s3Connector.readCsvFile(
        `templates/${cmoTemplate}`,
        myBucket
    );
    const combinedWorsheet = combinedWorkbook.addWorksheet(defaultSheetName);
    const workbookReader = new stream.xlsx.WorkbookReader(templateBody, {});
    for await (const worksheetReader of workbookReader) {
        let rowsInserted = 0;
        for await (const row of worksheetReader) {
            if (rowsInserted < contentStartingFromRow - 1) {
                // xls rows start with 1 not 0
                combinedWorsheet.addRow(row.values).commit();
            }
        }
    }
#  add other content to it.. 
```
or  for smaller set of data 

```
        await currentWorkbook.xlsx.load(templateBody);
        currentWorkSheet = this.currentWorkbook.worksheets[0];
#  loop and addRow to this worksheet
```

**Reply by vandungday** [reply]
thank you for greate solution.

In my case, I have large data (about gather 1 milion row) so I think using the your first solution to good work

**Comment by gomathims** [comment]
> 

Worksheet reader is not returned as per the order in Excel, if the excel is saved in MAC version of Excel. 
Suggestions?
